### ISSUE TITLE
Multi-GPU Support

### ISSUE THREAD

**Issue by 3cham** [benfred/implicit#366]
https://github.com/benfred/implicit/blob/42df436936a92fe64b065d9b7c3a9da1adfeb8a3/implicit/cuda/als.cu#L129

Hi, I find this library really useful and would like to have it supporting multiple GPUs since I have pretty big number of users. Any idea how could we extend this?

**Comment by 3cham** [comment]
Hi @benfred, I have ALS model with more than 40 mil. users and about 10k items with approx. 500 mil. ratings. The maximum memory I could get from one GPU is 16GB and we have two GPUs on each node. I tried to fit the model on one GPU but not successful. It would be great to split the computation on multiple GPUs. 

I have no experiences with CUDA but I think splitting the implicit matrix & its computation on multiple GPUs could be done without much changing current implementation (correct me if i'm wrong) & I would like to hear your opinion on this :) 

Actually I find `implicit` great when it comes to have a no-brainer but really fast ALS implementation. I currently use spark for this scale but it takes quite long time to train & predict.

**Comment by 3cham** [comment]
@benfred: yes, reducing the number of factor is what we did to fit it into our GPU. Having another GPU is a bit difficult since we rely on Google Cloud Platform. 

Off topic question: Could you please recommend any good online course for learning CUDA programming?

**Comment by tangzhy** [comment]
@benfred hi, I also have a strong demand for multi-gpu support.

1. OOM is my motivation to use multi-gpu like index_cpu_to_all_gpus() in faiss.
2. I'm using ALS right now.
3. Do you support specifying data type like np.float16 in gpu training, since that could halve memory usage significantly?
4. My data size: # of users = 159953032  / # of items = 151408354 / # of total interactions between them = 414543027.

Right now I'm stuck with my overwhelming data size which prevents me to train using gpu.
Any advice?

**Comment by fkurushin** [comment]
Hi @tangzhy 4 years passed) any updates on your problem?
